netbench/netbench-collector/src/main.rs

```diff
@@ -18,24 +36,363 @@ pub struct Args {
+        if self.as_server()? {
+            self.server_location.clone()
+        } else if self.as_client()? {
+            self.client_location.clone()
+        } else {
+            unimplemented!("Only --run-as server and --run-as client are supported options");
+        }
+    }
+    pub fn other_location(&self) -> Option<String> {
+        if self.as_server()? {
+            self.client_location.clone()
+        } else if self.as_client()? {
+            self.server_location.clone()
+        } else {
+            unimplemented!("Only --run-as server and --run-as client are supported options");
+        }
+    }
 }
 
-fn main() -> Result<()> {
-    let args = Args::from_args();
+pub trait RunHandle {
+    fn wait(self) -> Result<()>;
+    fn kill(self) -> Result<()>;
+}
+
+enum Handle {
+    Generic(GenericHandle),
+    Bpf(BpftraceHandle),
+}
 
-    // try to use bpftrace
-    if let Some(()) = bpftrace::try_run(&args)? {
-        return Ok(());
+impl RunHandle for Handle {
+    fn wait(self) -> Result<()> {
+        match self {
+            Self::Generic(h) => h.wait(),
+            Self::Bpf(h) => h.wait(),
+        }
     }
+    fn kill(self) -> Result<()> {
+        match self {
+            Self::Generic(h) => h.kill(),
+            Self::Bpf(h) => h.kill(),
+        }
+    }
+}
+
+fn run(args: Args) -> JoinHandle<Handle> {
+    tokio::spawn(async move {
+        // try to use bpftrace
+        if let Some(trace_handle) = bpftrace::try_run(&args).unwrap() {
+            return Handle::Bpf(trace_handle);
+        }
+
+        // fall back to the generic collector
+        Handle::Generic(generic::run(&args).unwrap())
+    })
+}
+
+#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq)]
+enum State {
+    NotReady,
+    Ready,
+    Running,
+    Finished,
+}
+
+#[derive(Serialize, Deserialize)]
+struct StateMessage {
+    version: String,
+    state: State,
+}
 
-    // fall back to the generic collector
-    generic::run(&args)?;
+impl From<State> for StateMessage {
+    fn from(value: State) -> StateMessage {
+        Self {
+            version: "18-04-2023".to_string(),
+            state: value,
+        }
+    }
+}
 
-    Ok(())
+impl TryFrom<u8> for State {
+    type Error = ();
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(State::NotReady),
+            1 => Ok(State::Ready),
+            2 => Ok(State::Running),
+            3 => Ok(State::Finished),
+            _ => Err(()),
+        }
+    }
+}
+
+impl From<State> for u8 {
+    fn from(value: State) -> u8 {
+        match value {
+            State::NotReady => 0,
+            State::Ready => 1,
+            State::Running => 2,
+            State::Finished => 3,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+struct StateTracker {
+    current_state: Arc<AtomicU8>,
+    location: String,
+    other_location: String,
+    verbose: bool,
+}
+
+macro_rules! log_polling {
```

Comment:
I think it would be better to link into a standard framework for this. Might be worth looking at [tracing](https://tracing.rs/tracing/#)